File: implement-trie-prefix-tree/implement-trie-prefix-tree.cpp
class Trie {
public:
    class Node{
    public:
        bool isWord;
        Node* children[26];

        Node() {
            isWord=false;
            for(int i=0;i<26;i++) children[i]=NULL;
        }
    };

    Node* root;

    Trie() {
        root = new Node();
    }
    
    void insert(string word) {
        Node* temp=root;
        for(char c : word)
        {
            if(temp->children[c-'a']==NULL) temp->children[c-'a']=new Node();
            temp=temp->children[c-'a'];
        }
        temp->isWord=true;
    }
    
    bool search(string word) {
        Node* temp=root;
        for(char c : word)
        {
            if(temp->children[c-'a']==NULL) return false;
            temp=temp->children[c-'a'];
        }
        return (temp->isWord);
    }
    
    bool startsWith(string prefix) {
        Node* temp=root;
        for(char c : prefix)
        {
            if(temp->children[c-'a']==NULL) return false;
            temp=temp->children[c-'a'];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.java
class Trie {

    private HashMap<Character,Trie> children;
    boolean isWord;

    public Trie() {
        children = new HashMap<>();
        isWord = false;
    }
    
    public void insert(String word) {
        Trie node = this;
        for(char c : word.toCharArray())
        {
            node.children.putIfAbsent(c,new Trie());
            node = node.children.get(c);
        }
        node.isWord = true;
    }
    
    public boolean search(String word) {
        Trie node = this;
        for(char c : word.toCharArray())
        {
            if(!node.children.containsKey(c)) return false;
            node = node.children.get(c);
        }
        return node.isWord == true;
    }
    
    public boolean startsWith(String prefix) {
        Trie node = this;
        for(char c : prefix.toCharArray())
        {
            if(!node.children.containsKey(c)) return false;
            node = node.children.get(c);
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */